pulsar-jms-tracing/src/main/java/com/datastax/oss/pulsar/jms/tracing/BrokerTracing.java

```diff
@@ -0,0 +1,553 @@
+
+    return enabledEvents;
+  }
+
+  @NotNull
+  private static TraceLevel getTraceLevel(PulsarService pulsar) {
+    String level =
+        pulsar
+            .getConfiguration()
+            .getProperties()
+            .getProperty("jmsTracingLevel", defaultTraceLevel.toString());
+    try {
+      return TraceLevel.valueOf(level.trim().toUpperCase());
+    } catch (IllegalArgumentException e) {
+      log.warn("Invalid tracing level: {}. Using default: {}", level, defaultTraceLevel);
+      return defaultTraceLevel;
+    }
+  }
+
+  private final LoadingCache<Subscription, TraceLevel> traceLevelForSubscription =
+      CacheBuilder.newBuilder()
+          .expireAfterWrite(cacheTraceLevelsDurationSec, TimeUnit.SECONDS)
+          .build(
+              new CacheLoader<Subscription, TraceLevel>() {
+                public TraceLevel load(Subscription sub) {
+                  Map<String, String> subProps = sub.getSubscriptionProperties();
+                  try {
+                    if (subProps == null || !subProps.containsKey("trace")) {
+                      PulsarAdmin admin =
+                          sub.getTopic().getBrokerService().getPulsar().getAdminClient();
+                      return BrokerTracing.getTracingLevel(admin, sub.getTopic());
+                    }
+
+                    return TraceLevel.valueOf(subProps.get("trace").trim().toUpperCase());
+                  } catch (IllegalArgumentException e) {
+                    log.warn(
+                        "Invalid tracing level: {}. Setting to NONE for subscription {}",
+                        subProps.get("trace"),
+                        sub);
+                    return TraceLevel.NONE;
+                  } catch (Throwable t) {
+                    log.error("Error getting tracing level", t);
+                    return TraceLevel.NONE;
+                  }
+                }
+              });
+
+  private final LoadingCache<Producer, TraceLevel> traceLevelForProducer =
+      CacheBuilder.newBuilder()
+          .expireAfterWrite(cacheTraceLevelsDurationSec, TimeUnit.SECONDS)
+          .build(
+              new CacheLoader<Producer, TraceLevel>() {
+                public TraceLevel load(Producer producer) {
+                  try {
+                    PulsarAdmin admin =
+                        producer.getCnx().getBrokerService().getPulsar().getAdminClient();
+                    Topic topic = producer.getTopic();
+
+                    return BrokerTracing.getTracingLevel(admin, topic);
+                  } catch (Throwable t) {
+                    log.error("Error getting tracing level", t);
+                    return TraceLevel.NONE;
+                  }
+                }
+              });
+
+  @NotNull
+  private static TraceLevel getTracingLevel(PulsarAdmin admin, Topic topic) {
+    Map<String, String> props = null;
+    try {
+      props =
+          admin.topics().getProperties(TopicName.get(topic.getName()).getPartitionedTopicName());
```

Comment:
We are executing this code in the hot paths, so couple of comments:
- it may add latency spikes
- it may fail with some temporary error
- it may introduce side effects during topic loading (all the weird combination of events that happen while loading a topic)

I don't think we can do much to mitigate these problems, but I suggest to log one line before calling xx.getProperties() this way in case of latency spikes or other problems, it is possible to see in the logs that a blocking operation is started.